### PR TITLE
planner: support using expression index when conditions are connected by `or` or `not` 

### LINKED ISSUE #25961: expression index can not be used if condition connected by or
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

``` sql
create table t(a int, b varchar(10), key((lower(b)), (a+1)));
desc  select  * from t use index(expression_index) where (lower(b) = "a" and a+1 = 2) or (lower(b) = "b" and a+1 = 3);
```

### 2. What did you expect to see? (Required)

``` SQL
mysql> desc  select  * from t use index(expression_index) where (lower(b) = "a" and a+1 = 2) or (lower(b) = "b" and a+1 = 3);
+---------------------------------+---------+-----------+------------------------------------------------------+--------------------------------------------------------------------+
| id                              | estRows | task      | access object                                        | operator info                                                      |
+---------------------------------+---------+-----------+------------------------------------------------------+--------------------------------------------------------------------+
| Projection_4                    | 0.20    | root      |                                                      | test.t.a, test.t.b                                                 |
| └─IndexLookUp_7                 | 0.20    | root      |                                                      |                                                                    |
|   ├─IndexRangeScan_5(Build)     | 0.20    | cop[tikv] | table:t, index:expression_index(lower(`b`), `a` + 1) | range:["a" 2,"a" 2], ["b" 3,"b" 3], keep order:false, stats:pseudo |
|   └─TableRowIDScan_6(Probe)     | 0.20    | cop[tikv] | table:t                                              | keep order:false, stats:pseudo                                     |
+---------------------------------+---------+-----------+------------------------------------------------------+--------------------------------------------------------------------+
4 rows in set (0.00 sec)
```

### 3. What did you see instead (Required)
```
mysql> desc  select  * from t use index(expression_index) where (lower(b) = "a" and a+1 = 2) or (lower(b) = "b" and a+1 = 3);
+---------------------------------+----------+-----------+------------------------------------------------------+----------------------------------------------------------------------------------------------------------------------+
| id                              | estRows  | task      | access object                                        | operator info                                                                                                        |
+---------------------------------+----------+-----------+------------------------------------------------------+----------------------------------------------------------------------------------------------------------------------+
| Selection_5                     | 8000.00  | root      |                                                      | or(and(eq(lower(test.t.b), "a"), eq(plus(test.t.a, 1), 2)), and(eq(lower(test.t.b), "b"), eq(plus(test.t.a, 1), 3))) |
| └─IndexLookUp_8                 | 10000.00 | root      |                                                      |                                                                                                                      |
|   ├─IndexFullScan_6(Build)      | 10000.00 | cop[tikv] | table:t, index:expression_index(lower(`b`), `a` + 1) | keep order:false, stats:pseudo                                                                                       |
|   └─TableRowIDScan_7(Probe)     | 10000.00 | cop[tikv] | table:t                                              | keep order:false, stats:pseudo                                                                                       |
+---------------------------------+----------+-----------+------------------------------------------------------+----------------------------------------------------------------------------------------------------------------------+
4 rows in set, 1 warning (0.00 sec)
```

### 4. What is your TiDB version? (Required)
master

### PR DESCRIPTION
**Pull request by wjhuang2016**
Signed-off-by: wjhuang2016 <huangwenjun1997@gmail.com>

### What problem does this PR solve?

Issue Number: close #25961 

Problem Summary:
Previously we didn't support `or` or `not`. 

### What is changed and how it works?

Small refractory.

### Check List 

Tests 

- Unit test

Side effects

### Release note 

- No release note

### CODE REVIEW

**Comment by jackysp** [comment]
No release note?

**Comment by wjhuang2016** [comment]
> No release note?

No

File: cmd/explaintest/r/explain_generate_column_substitute.result [hunk 1]
```diff
@@ -414,3 +414,18 @@ StreamAgg	1.00	root		funcs:count(Column#6)->Column#4
...
+drop table if exists t;
+create table t(a int, b varchar(10), key((lower(b)), (a+1)));
+insert into t values (1, "A"), (2, "B"), (3, "C"), (4, "D"), (5, "E"), (6, "F");
+desc format = 'brief' select * from t where (lower(b) = "a" and a+1 = 2) or (lower(b) = "b" and a+1 = 5);
+id	estRows	task	access object	operator info
+Projection	0.20	root		test.t.a, test.t.b
+└─IndexLookUp	0.20	root		
+  ├─IndexRangeScan(Build)	0.20	cop[tikv]	table:t, index:expression_index(lower(`b`), `a` + 1)	range:["a" 2,"a" 2], ["b" 5,"b" 5], keep order:false, stats:pseudo
+  └─TableRowIDScan(Probe)	0.20	cop[tikv]	table:t	keep order:false, stats:pseudo
+desc format = 'brief' select * from t where not (lower(b) >= "a");
```
Comment (Reminiscent) [inline cmd/explaintest/r/explain_generate_column_substitute.result:426]: Can the expression index be used here?

Comment (wjhuang2016) [reply cmd/explaintest/r/explain_generate_column_substitute.result:427]: I add a `analyze table t` to fix it.

**Review by winoros — APPROVED** [review]

**Review by Reminiscent — APPROVED** [review]

**Pull request merged**